### ISSUE TITLE
About the documentation.

### ISSUE THREAD

**Issue by jashika-oss** [oracle/tribuo#47]
https://tribuo.org/learn/4.0/docs/

Isn't the documentation and implementation code listed in the above URL wrong?

For example, the variable definition is 'irisData', but the variable reference says 'irisesSource'. This will result in a compilation error.

Here's what else is going on

Constructor calls that don't have a LibSVMDataSource, for example.
https://qiita.com/jashika/items/aa8ca340deb81a59cbea

A call to a method that does not exist. The class name to be called is likely to be different, etc.
https://qiita.com/jashika/items/1ba0cad613ec919adaa7

**Comment by Craigacp** [comment]
Ooops, yes there a couple of bugs there. That'll show me for writing the code snippets directly in Markdown without an IDE. I've updated the website.

The file loading issue with bezdekIris.data is because you need Tribuo 4.0.1 which makes the CSV parser slightly more relaxed about unnecessary newlines. I had previously deleted that newline from the version on my laptop so I didn't notice when prepping the tutorials. I bumped the version specified in the quickstart to 4.0.1, we missed that when deploying the website.

Unfortunately I don't read Japanese, and I think Google Translate is making a mess of the translation, so I'm not sure what other issues you are raising. It looks like there's something about null handling in the LibSVMDataSource? The postConfig method is used to validate the either/or nature of the url and path in the data source when it's loaded via our configuration system, and is not required to be called by the standard constructors. I guess we should have stronger null handling in the constructor, but an NPE is likely to be developer error, and it's pretty obvious what causes one in that code path. I'll add a note to fix that handling when we do a pass through.

**Comment by jashika-oss** [comment]
Thanks

I'm a not native English speaker, There may be a slight misrepresentation.

Regarding null, when the java.net.URL class is null, the error occurs for the first time at the following #Error Location after the processing of org.tribuo.datasource.LibSVMDataSource#read is done.

#Error Location.
try (BufferedReader r = new BufferedReader(new InputStreamReader(url.openStream(),StandardCharsets.UTF_8))) {

Even if you don't go that far into the process, you know it will be an error, so it would be better for the code to make it an error when it is discovered that the error has been confirmed.

org.tribuo.datasource.LibSVMDataSource#postConfig's null check looks like the following #Corresponding Place, but if you insert null in the java.nio.file.Path, it throw null pointer exception by constructor. if you insert null in the java.net.URL,
throw null pointer exception by org.tribuo.datasource. LibSVMDataSource#read method.

This class  haven't setter and getter. #The relevant location would be an impossible conditional expression, wouldn't it?

#Corresponding Place
if ((url == null) && (path == null)) {

Also, there is no mention of javadoc around the relevant part, so the design intent is unreadable. It says something like this.

**Comment by jashika-oss** [comment]
I would like to add one more thing.
Considering edge devices and java, I think you need to do the math in FP16.

**Comment by Craigacp** [comment]
Ok, we can fix the javadoc on the constructors. The null check can be promoted into the private constructor. The main reason for the null check in the postConfig is to validate that either the path or the url are set, and not both. That's not possible with the regular constructors, at most one of them is non-null, but it's entirely reasonable to want the constructor to complain if it's passed a null.

The class intentionally doesn't have setters, as it's effectively immutable to make the provenance information easier to compute.

**Comment by Craigacp** [comment]
> I would like to add one more thing.
> Considering edge devices and java, I think you need to do the math in FP16.

The Java language doesn't currently support an fp16 type, and there is minimal hardware support for it on CPUs. You can use fp16 GPU hardware via the ONNX or TensorFlow interfaces assuming that the ONNX or TF model is configured appropriately.

What kind of edge devices do you have in mind?

**Comment by jashika-oss** [comment]
Sorry if I misunderstood your meaning.

<blockquote>
The class intentionally doesn't have setters, as it's effectively immutable to make the provenance information easier to compute.
</blockquote>

I know it. But I wanna say it have check logic 'if ((url == null) && (path == null)) ' on line 137.

The class can't make instance with url == null && path == null (Except for the case of using refrection to make it null). Because it throw exception. And I can't set null in url and path. Because it haven't setter and getter.

I think 1-3.

1 It's dead logic on the class. Because It can't make instance looks like this.
2 It's dead logic on the class. Because It set data Just constractor. this(path,path.toUri().toURL(),outputFactory,false,false,0) 
3 It's dead logic on the class. Because You said 'It's effectively immutable'. I think too. It's effectively immutable but is it set again?

public void postConfig() throws IOException {
                                                               ・
                                                               ・
                                                               ・
1--->if ((url == null) && (path == null)) {
            throw new PropertyException("","path","At most one of url and path must be set.");
2--->} else if ((url != null) && (path != null) && !path.toUri().toURL().equals(url)) {
            throw new PropertyException("","path","At most one of url and path must be set");
         } else if (path != null) {
            // url is the store of record.
            try {
3--->       url = path.toUri().toURL();
            } catch (MalformedURLException e) {
                throw new PropertyException(e,"","path","Path was not a valid URL");
            }
        }
        read();
    }


<blockquote>
What kind of edge devices do you have in mind?
</blockquote>

The only thing I have at the moment is an android phone and iphone.

I think don't must need FP16 right now. But I want to it as soon as possible.
it could be something like google grass, or it could be a bit more advanced and be like a contact lens. I thought it would be interesting to be able to support devices that don't exist right now, not necessarily right away, but that will be in regular use 10 years from now. When a small, lightweight device comes out at a price that anyone can afford, even if the accuracy is reduced to FP16, there will be requirements to embed it, so I think it would be nice if Java could support it eventually.


As for tenser-frow, there's also a java front end coming out, and I don't want it to be lose. And I think there is value in being able to complete it all in Java. In reality, it cannot be completed using only Java because of OpenCL or Cuda, I think it is worth aiming for.

**Comment by Craigacp** [comment]
Ah, ok. You should never call `postConfig` directly, it's a public method because it's used by Tribuo's configuration system OLCUT (https://github.com/oracle/olcut/blob/develop/README-Configuration.md), and is part of the `Configurable` interface which denotes that things can be managed by that configuration system. OLCUT constructs objects out of configuration files (see the configuration tutorial here - https://github.com/oracle/tribuo/blob/main/tutorials/configuration-tribuo-v4.ipynb), first the private no-args constructor `private LibSVMDataSource()` is called, then the field values are updated via reflection (assigning to the ones marked `@Config`), then `postConfig()` is called so the class can validate that it's invariants are satisfied. That's why the logic is there, and all three cases are possible in objects built via configuration as neither `path` or `url` are marked mandatory unlike `outputFactory`. 

I've added the null checks to LibSVMDataSource and updated the javadoc in this commit - https://github.com/oracle/tribuo/commit/defd11330b24f0e489972e150a6de18c19028193, there will be a PR that includes it along with several other documentation fixes at some point this week.

In the case of TensorFlow Java, I'm one of the main contributors to that project as well, and Tribuo will track the new releases from the TF JVM SIG as they arrive. It's already possible using Tribuo and the current TensorFlow interface to load in TF models which use fp16 operations, and have the TF native code use fp16 HW in GPUs to execute it. Similarly it's currently possible to compile the ONNX Runtime (which provides Tribuo's ONNX model support) to execute on fp16 GPU hardware by using the onnxruntime_gpu artifact (https://search.maven.org/artifact/com.microsoft.onnxruntime/onnxruntime_gpu/1.4.0/jar) as a drop in replacement for the onnxruntime artifact (https://search.maven.org/artifact/com.microsoft.onnxruntime/onnxruntime/1.4.0/jar) that Tribuo depends on. The JVM already supports aarch64, and I'm sure that TF Java and ONNX Runtime's Java API will support it soon.

**Comment by jashika-oss** [comment]
Thank you very much, your time for me.

I was worried that my English would not be able to convey what I wanted to say. 

I said 'I don't understand the design intent. Because it have not javadoc'. It's a public class, public method, and non internal package. but non javadoc. I thought internal logic. But I couldn't found document.

I was looking for looks like this answer.

<blockquote>
Ah, ok. You should never call postConfig directly, it's a public method because it's used by Tribuo's configuration system OLCUT 
</blockquote>

Is the other source code intended to do the same?
・public class, public method, non internal package non javadoc.

<blockquote>
In the case of TensorFlow Java, I'm one of the main contributors to that project as well, and Tribuo will track the new releases from the TF JVM SIG as they arrive.
</blockquote>

Thank you. I'll try to implement it a bit and run it.

**Comment by jashika-oss** [comment]
I'm sorry. It's very small. 
But I think important for first time user. 
I found small problem on document.

https://tribuo.org/learn/4.0/docs/

```
1. var trainSet = new MutableDataset<>(new LibSVMDataSource(Paths.get("train-data"),new LabelFactory()));
2. var model    = new LogisticRegressionTrainer().train(trainSet);
3. var eval     = new LabelEvaluator()
                        .evaluate(new LibSVMDataSource(Paths.get("test-data"),trainSet.getOutputFactory()));
```

Is it correct line 3? The method must need model (The model is line 2 instance).
```
var eval = new LabelEvaluator()
                .evaluate(model, new LibSVMDataSource(Paths.get("test-data"), trainSet.getOutputFactory()));
```

**Comment by Craigacp** [comment]
Yep, I'll fix that too.

The `postConfig` method is the only one used by the configuration system. The provenance objects all require a public constructor which accepts a single `Map<String,Provenance>` argument, as part of their serialization mechanism. I'll put some javadoc in for all the `postConfig` methods noting that they are part of the configuration system and should not be called by user code.

**Comment by Craigacp** [comment]
Fixed the quickstart bit here - https://github.com/oracle/tribuo-site/commit/defced4cebf6dd2b8dee07174778e22b15719471

**Comment by jashika-oss** [comment]
Thank you.

I have a few other ideas. But I haven't been able to read all the source code yet. I'll talk about it when I understand it a bit better.

Somehow, overall, I'm starting to understand what the source code is trying to do. But I haven't fully experimented with it yet, so I'm going to try to move it around some more.

At first, I was going to just talk about the documentation. Thank you for teaching me so much.
